src/compiler/factory.ts

```diff
@@ -19,21 +22,16 @@ namespace ts {
     }
 
     /* @internal */ export function createNodeArray<T extends Node>(elements?: T[], hasTrailingComma?: boolean): MutableNodeArray<T>;
-    export function createNodeArray<T extends Node>(elements?: ReadonlyArray<T>, hasTrailingComma?: boolean): NodeArray<T>;
+    export function createNodeArray<T extends Node>(elements?: Nodes<T>, hasTrailingComma?: boolean): NodeArray<T>;
     /**
      * Make `elements` into a `NodeArray<T>`. If `elements` is `undefined`, returns an empty `NodeArray<T>`.
      */
-    export function createNodeArray<T extends Node>(elements?: ReadonlyArray<T>, hasTrailingComma?: boolean): NodeArray<T> {
-        if (elements) {
-            if (isNodeArray(elements)) {
-                return elements;
-            }
-        }
-        else {
-            elements = [];
+    export function createNodeArray<T extends Node>(elements: Nodes<T> = [], hasTrailingComma?: boolean): NodeArray<T> {
+        if (isNodeArray(elements)) {
+            return elements;
         }
 
-        const array = <NodeArray<T>>elements;
+        const array = elements as MutableNodeArray<T>;

```

Comment:
Perhaps, if its not already a `NodeArray`, we should instead clone it via `.slice()` before modifying it? Then we wouldn't need the `Nodes<T>` type, though at first glance I'm not certain what the memory/performance impact would be.

Reply:
I tested out adding `slice` inside of `createNodeArray` (at the bottom of the `if (elements)` block), and actually saw a slight decrease in emit time for some reason -- might have to do with the fact that `createNodeArray` would otherwise have to change the shape of an array object acquired from elsewhere, but if we create a fresh slice it may be allocated with the correct shape already.

The most efficient option may be to avoid ever calling `createNodeArray` on an existing array, and instead always start by calling `createNodeArray` to get an empty mutable array, then pushing to it.